crates/precompile-evm-balances-erc20/src/lib.rs

```diff
@@ -0,0 +1,410 @@
+//! A precompile to interact with `pallet_evm_balances` instances using the ERC20 interface standard.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+    sp_runtime::{self, traits::CheckedSub},
+    sp_std::{marker::PhantomData, prelude::*},
+    storage::types::StorageDoubleMap,
+    traits::{tokens::currency::Currency, StorageInstance},
+    Blake2_128Concat,
+};
+use pallet_evm::{
+    ExitError, Precompile, PrecompileFailure, PrecompileHandle, PrecompileOutput, PrecompileResult,
+};
+use precompile_utils::{
+    keccak256, revert, succeed, Address, Bytes, EvmDataReader, EvmDataWriter, EvmResult, LogExt,
+    LogsBuilder, PrecompileHandleExt,
+};
+use sp_core::{Get, H160, U256};
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+/// Solidity selector of the Transfer log, which is the Keccak of the Log signature.
+pub const SELECTOR_LOG_TRANSFER: [u8; 32] = keccak256!("Transfer(address,address,uint256)");
+
+/// Solidity selector of the Approval log, which is the Keccak of the Log signature.
+pub const SELECTOR_LOG_APPROVAL: [u8; 32] = keccak256!("Approval(address,address,uint256)");
+
+/// Metadata of an ERC20 token.
+pub trait Erc20Metadata {
+    /// Returns the name of the token.
+    fn name() -> &'static str;
+
+    /// Returns the symbol of the token.
+    fn symbol() -> &'static str;
+
+    /// Returns the decimals places of the token.
+    fn decimals() -> u8;
+}
+
+/// Approves storage instance to manage the allowance related data.
+pub struct Approves;
+
+impl StorageInstance for Approves {
+    const STORAGE_PREFIX: &'static str = "Approves";
+
+    fn pallet_prefix() -> &'static str {
+        "EvmBalancesErc20Approves"
+    }
+}
+
+/// Storage type used to store approvals, since `pallet_evm_balances` doesn't handle this behavior.
+/// (Owner => Allowed => Amount).
+pub type ApprovesStorage<Runtime> = StorageDoubleMap<
```

Comment:
Can we have a pallet for this instead? It will sure come in handy for a more advanced management of the state, like upgrades and etc, if we need it later.

We can call it `pallet_erc20_approvals` or `pallet_erc20`

Reply:
Sure, `pallet-token-wrapper` has been added.